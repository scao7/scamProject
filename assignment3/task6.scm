(define (sdisplay n z)
    (define (iter x m)
        (if (= m 0)  
                (print "...")
            (begin
                (print (stream-car x) ",")
                (iter (stream-cdr x) (- m 1)) 
            )
        )
    )   
    (print "(")
    (iter z n)
    (print ")")
)  

(define (quad a b c step)
    (define (square x) (* x x))
    (define (get-f val) (+ (* a (square val)) (* b val) c))
    (define (stepper s)
        (cons-stream (+ (stream-car s) (real step)) (stepper (stream-cdr s)))
        )
    (define (f-stream s)
        (cons-stream (get-f (stream-car s)) (f-stream (stream-cdr s)))
        )
    (define x (cons-stream (real 0.0) (stepper x)))
    (define f (cons-stream (get-f (stream-car x)) (f-stream (stream-cdr x))))
)


(define (integrate s height)
    (define (trapzoid top base)(* (/ (+ top base) 2.0) height))
    (define (integ-help st self)
        (cons-stream (+ (stream-car self) (trapzoid (stream-car st) (stream-car (stream-cdr st)))) (integ-help (stream-cdr st) (stream-cdr self)))
        )
    (define integ (cons-stream (real 0.0) (integ-help s integ)))
)


(define (derivate s step f0)
    (define (undo integ top)
        (- (* (/ integ step) 2.0) top)
     )
    (define (deriv s step top)
        (let ((trap (- (stream-car (stream-cdr s)) (stream-car s))))
            (define base (undo trap top))
            (cons-stream base (deriv (stream-cdr s) step base))
            )
        )
    (cons-stream (real f0) (deriv s step f0))
    )

(define (same-stream? s1 s2 n t)
    (cond
        ((= n 0) #t)
        ((<= (abs (- (stream-car s1) (stream-car s2))) t) (same-stream? (stream-cdr s1) (stream-cdr s2) (- n 1) t))
        (else #f)
        )
)


(define (main)
    (define env this)
    (define (iter expr)
         (if (not (eof?)) (begin (eval expr env) (iter (readExpr))))
         )
    (setPort (open (getElement ScamArgs 1) 'read))
    (iter (readExpr))
    )
