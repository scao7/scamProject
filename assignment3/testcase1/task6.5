(define qs (quad 1 2 1 0.0109890110))
(define is (integrate qs 0.0109890110))
(define ds (derivate is 0.0109890110 (stream-car qs)))
(sdisplay 5 qs) (println)
(sdisplay 5 is) (println)
(sdisplay 5 ds) (println)
(inspect (same-stream? qs ds 10 .001))
