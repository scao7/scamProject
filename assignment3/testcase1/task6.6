(define qs (quad 0 -3 1 0.0833333333))
(define is (integrate qs 0.0833333333))
(define ds (derivate is 0.0833333333 (stream-car qs)))
(sdisplay 5 qs) (println)
(sdisplay 5 is) (println)
(sdisplay 5 ds) (println)
(inspect (same-stream? qs ds 10 .001))
