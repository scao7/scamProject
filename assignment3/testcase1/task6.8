(define qs (quad 2 3 1 0.0192307692))
(define is (integrate qs 0.0192307692))
(define ds (derivate is 0.0192307692 (stream-car qs)))
(sdisplay 5 qs) (println)
(sdisplay 5 is) (println)
(sdisplay 5 ds) (println)
(inspect (same-stream? qs ds 10 .001))
