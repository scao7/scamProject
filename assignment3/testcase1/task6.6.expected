(1.0000000000,0.7500000001,0.5000000002,0.2500000003,3.999999e-10,...)
(0.000000e+00,0.0729166666,0.1250000000,0.1562500000,0.1666666667,...)
(1.0000000000,0.7500000001,0.5000000002,0.2500000003,4.000003e-10,...)
(same-stream? qs ds 10 0.0010000000) is #t
